### PR TITLE
use wchar_t for all potentially-wide-character handling

### LINKED ISSUE #14161: ARM: patch for -Werror=sign-compare
Arm assumes that char types are unsigned while x86 assumes they are signed.  This creates problems with comparisons.  Specifically

    src/wcwidth.h: In function ‘int bisearch(wchar_t, const interval*, int)’:
    src/wcwidth.h:76:11: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
       if (ucs < table[0].first || ucs > table[max].last)

The file `src/wcwidth.h` should have the interval struct changed to be

    struct interval {
      wchar_t first;
      wchar_t last;
    };

Tested on armv7, i686 and x86_64.  This issue is also present in the 0.C release, and is the only ARM issue I could find.

### LINKED ISSUE #14540: Can't compile on raspberry pi 2
CCACHE_CPP2=1 ccache clang++  -DRELEASE -DGIT_VERSION -DLOCALIZE -ffast-math -Os -Wall -Wextra   -Werror --std=c++11 -MMD -I/usr/include/lua5.2  -DLUA -I/usr/include/ncursesw -c src/catacharset.cpp -o obj/catacharset.o
In file included from src/catacharset.cpp:5:
src/wcwidth.h:76:11: error: comparison of integers of different signs: 'wchar_t' and 'const int' [-Werror,-Wsign-compare]
  if (ucs < table[0].first || ucs > table[max].last)
      ~~~ ^ ~~~~~~~~~~~~~~
src/wcwidth.h:76:35: error: comparison of integers of different signs: 'wchar_t' and 'const int' [-Werror,-Wsign-compare]
  if (ucs < table[0].first || ucs > table[max].last)
                              ~~~ ^ ~~~~~~~~~~~~~~~
src/wcwidth.h:80:13: error: comparison of integers of different signs: 'wchar_t' and 'const int' [-Werror,-Wsign-compare]
    if (ucs > table[mid].last)
        ~~~ ^ ~~~~~~~~~~~~~~~
src/wcwidth.h:82:18: error: comparison of integers of different signs: 'wchar_t' and 'const int' [-Werror,-Wsign-compare]
    else if (ucs < table[mid].first)
             ~~~ ^ ~~~~~~~~~~~~~~~~
4 errors generated.
Makefile:541: recipe for target 'obj/catacharset.o' failed
make: *** [obj/catacharset.o] Error 1
make: *** Waiting for unfinished jobs....

### PR DESCRIPTION
**Pull request by sparr**
I'm not too savvy on the ways of wide character handling, so this patch definitely needs some scrutiny. I started out just trying to fix #14540 and the rabbit hole got a bit deeper. Existing code in various places used `unsigned`, `long`, `int`, and `uint32_t` to store wide characters. In addition to not necessarily being equivalent, some or all of those are incompatible with platforms where wchar_t is signed (since we use `-Werror -Wsign-compare`).

Fixes #14161

### CODE REVIEW

File: src/cata_tiles.cpp [hunk 1]
```diff
@@ -1125,7 +1125,7 @@ bool cata_tiles::draw_from_id_string(std::string id, TILE_CATEGORY category,
     }
 
     if (it == tile_ids.end()) {
-        long sym = -1;
+        wchar_t sym = -1;
```
Comment (kevingranade) [inline src/cata_tiles.cpp]: If the compiler warned about ```wchar_t < int```, I'm pretty sure this is going to make it angry too.

**Comment by kevingranade** [comment]
Instead of wchar_t (which is poorly-defined and deprecated for every use except interacting with windows APIs), I think uint32_t is the best choice, it will definitely handle every unicode codepoint.

File: src/cata_tiles.cpp (hunk 1 above)
Comment (kevingranade) [inline src/cata_tiles.cpp]: In other words, if you want to make this unsigned, the method needs to be refactored to not try and assign a negative number to it.

Comment (sparr) [inline src/cata_tiles.cpp]: the warning in question is for comparisons. I don't think there's an equivalent warning for assignments. implicit casting should function appropriately here. further down sym is only used if it's >0 and <256 so the only way -1 would be a problem here is if it wchar_t is an unsigned byte, and we've got much bigger problems if that happens.

**Comment by sparr** [comment]
If uint32_t will work, I can redo this for that type. I just used wchar_t because it was the one being used where the bug was spotted.

File: src/cata_tiles.cpp (hunk 1 above)
Comment (kevingranade) [inline src/cata_tiles.cpp]: If it's not a matter of "will it work", it's still a matter of not behaving as expected, on some systems sym is initialized to -1, on some it's initialized to 65535, and on others it's initialized to 4294967295, that's asking for trouble :)

Comment (sparr) [inline src/cata_tiles.cpp]: oh, if that's the only concern I'd switch the initialization here to UNKNOWN_UNICODE

Comment (kevingranade) [inline src/cata_tiles.cpp]: The negative number was not the only concern, wchar_t is a bad idea.

**Comment by sparr** [comment]
@kevingranade should I go for the minimum replacement of wchar_t with uint32_t to cover the cases already covered by this PR, or should I aim to more broadly extricate wchar_t? If the latter, how do you want to handle the templated to_string() in compatibility.h that's supposed to be able to convert integers to strings but should be disabled for wide characters?

**Comment by kevingranade** [comment]
Whatever you feel like tackling is fine, the only immediate need is
silencing the warnings.
I'd have to look more closely at the code, but "convert integers to strings
but should be disabled for wide characters" doesn't make any sense, wchar_t
is an integer.

**Pull request merged**